kibana/connecting_client_test.go

```diff
@@ -0,0 +1,137 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package kibana
+
+import (
+	"io/ioutil"
+	"net/http"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/kibana"
+
+	"github.com/elastic/apm-server/convert"
+)
+
+func TestNewConnectingClientFrom(t *testing.T) {
+	c := NewConnectingClient(mockCfg)
+	require.NotNil(t, c)
+	assert.Nil(t, c.(*ConnectingClient).client)
+	assert.Equal(t, mockCfg, c.(*ConnectingClient).cfg)
+}
+
+func TestConnectingClient_Send(t *testing.T) {
+	t.Run("Send", func(t *testing.T) {
+		c := mockClient()
+		r, err := c.Send(http.MethodGet, "", nil, nil, nil)
+		require.NoError(t, err)
+		assert.Equal(t, mockBody, r.Body)
+		assert.Equal(t, mockStatus, r.StatusCode)
+	})
+
+	t.Run("SendError", func(t *testing.T) {
+		c := NewConnectingClient(mockCfg)
+		r, err := c.Send(http.MethodGet, "", nil, nil, nil)
+		require.Error(t, err)
+		assert.Equal(t, err, errNotConnected)
+		assert.Nil(t, r)
+	})
+}
+
+func TestConnectingClient_GetVersion(t *testing.T) {
+	t.Run("GetVersion", func(t *testing.T) {
+		c := mockClient()
+		v, err := c.GetVersion()
+		require.NoError(t, err)
+		assert.Equal(t, mockVersion, v)
+	})
+
+	t.Run("GetVersionError", func(t *testing.T) {
+		c := NewConnectingClient(mockCfg)
+		v, err := c.GetVersion()
+		require.Error(t, err)
+		assert.Equal(t, err, errNotConnected)
+		assert.Equal(t, common.Version{}, v)
+	})
+}
+
+func TestConnectingClient_SupportsVersion(t *testing.T) {
+	t.Run("SupportsVersionTrue", func(t *testing.T) {
+		c := mockClient()
+		s, err := c.SupportsVersion(common.MustNewVersion("7.3.0"))
+		require.NoError(t, err)
+		assert.True(t, s)
+	})
+	t.Run("SupportsVersionFalse", func(t *testing.T) {
+		c := mockClient()
+		s, err := c.SupportsVersion(common.MustNewVersion("7.4.0"))
+		require.NoError(t, err)
+		assert.False(t, s)
+	})
+
+	t.Run("SupportsVersionError", func(t *testing.T) {
+		c := NewConnectingClient(mockCfg)
+		s, err := c.SupportsVersion(common.MustNewVersion("7.3.0"))
+		require.Error(t, err)
+		assert.Equal(t, err, errNotConnected)
+		assert.False(t, s)
+	})
+}
+
+func TestConnectingClient_Connected(t *testing.T) {
+	t.Run("Connected", func(t *testing.T) {
+		c := mockClient()
+		require.True(t, c.Connected())
+	})
+
+	t.Run("NotConnected", func(t *testing.T) {
+		c := NewConnectingClient(mockCfg)
+		require.False(t, c.Connected())
+	})
+}
+
+type rt struct {
+	resp *http.Response
+}
+
+var (
+	mockCfg     = common.MustNewConfigFrom(`{"enabled": "false", "host": "non-existing"}`)
+	mockBody    = ioutil.NopCloser(convert.ToReader(`{"response": "ok"}`))
```

Comment:
this is the first I've seen `ToReader` - it should probably just return a `ioutil.NopCloser` since we're wrapping that everywhere it's used.

Reply:
I only moved this, so didn't check in detail. But there are other places where we don't wrap it e.g. [agentcfg/fetch.go#L63](https://github.com/elastic/apm-server/blob/master/agentcfg/fetch.go#L63).